### PR TITLE
[flutter_tools] build ios-frameworks: option to exclude plugin frameworks from the build

### LINKED ISSUE #114692: [Add2App] provide flag for 'build ios-framework' command to build only App.xcframework
Feature request: provide flag for `flutter build ios-framework` command to build only `App.xcframework`  
or to prevent building of third-party pod dependencies as frameworks

## Use case

I have a Flutter module (let's say `FlutterModule`) integrated to an iOS app (`IOSApp`) using frameworks. My setup is similar to [Option C](https://docs.flutter.dev/development/add-to-app/ios/project-setup#option-c---embed-application-and-plugin-frameworks-in-xcode-and-flutter-framework-with-cocoapods) from the add-to-app doc for iOS (with an improvement of embedding application and plugin frameworks to XCode automatically using CocoaPods).  
- `FlutterModule` depends on third-party Flutter plugins, some of them have iOS dependencies, e.g. `sentry_flutter` which [depends](https://github.com/getsentry/sentry-dart/blob/7ade5af85bc00839aa6788a05725d7f9ecda0b6d/flutter/ios/sentry_flutter.podspec) on `Sentry` pod. 
- `IOSApp` also uses `CocoaPods` to add third-party iOS dependencies.

The problem appears when I want to add to `IOSApp`'s Podfile a dependency that reference another dependency that is referenced by Flutter plugin. This can be hard to understand, so here is the schema:
`IOSApp` -> `Pod1` -> `Pod2`
`IOSApp` -> `flutter_plugin.xcframework` -> `Pod2.xcframework`

So my `IOSApp` and `FlutterModule` have the same transitive platform dependency.

Because `Pod2` is included twice, first through `CocoaPods` and second directly as a framework, this causes multiple `Redefinition of` issues in XCode.

Sentry example

>  Note that behavior will be the same for case
`IOSApp` -> `Pod2`
`IOSApp` -> `flutter_plugin.xcframework` -> `Pod2.xcframework`

Currently `flutter build ios-framework` command produces following outputs for the build mode:
- `App.xcframework` (desired, compiled Dart code)
- `Flutter.xcframework` (`Flutter.podspec` in my case, if `--cocoapods` flag is provided. Desired)
- `FlutterPluginRegistrant.xcframework` (desired, registration code for plugins)
- `flutter_plugin.xcframework` (desired, each plugin is a separate framework)
- `Pod2.xcframework` (If plugin has dependencies in its `podspec`, frameworks will be built for each of them. Undesired: `Pod2` can be also included  to iOS app as a dependency, causing issues) 

---
I didn't create this as an issue because obviously I can just remove the duplicated frameworks from the Flutter command output to not include them. UPD: submit an issue https://github.com/flutter/flutter/issues/130220 because this workaround is not working for some plugins, like `datadog_flutter_plugin`

But this solution is not universal. Some downsides are:
- each time I add new dependency to `FlutterModule` I have to make sure it doesn't include CocoaPods dependency that conflicts with CocoaPods dependency of `IOSApp`. (Note that `Sentry` pod wasn't the only that caused issues)
- as a result of the previous I can't safely publish my `FlutterModule` itself as a CocoaPods dependency to be consumed by any iOS apps
- because `flutter build ios-framework` produces frameworks that will be removed right after the creation, this has a significant impact on build time. We can speed up this command by preventing the build of undesired frameworks.

## Proposal

Modify `build ios-framework` command to support this use case. I thought about two options:

- **Option 1**: flag to prevent third-party pod dependencies to be built as frameworks.  
Current solution is to build every target from `.ios/Pods/Pods.xcodeproj` (see [_producePlugins](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/commands/build_ios_framework.dart#L475) function). Targets include both Flutter plugins and their pod dependencies. We can try to change this, but I am not sure if it is possible to build only selected targets without building their dependencies.
Another question here is how to then easily include all of the remaining pods dependencies to app. Probably we can combine this with `--cocoapods` flag and generate another `podspec` with all third-party pods listed.
- **Option 2**: flag to build only `App.xcframework`
All we need to do is to prevent `_producePlugins` function from running. Inclusion of the pods dependencies to the app will be developer's responsibility and not the Flutter's (unlike in **Option 1**, it's easy to do)

I personally implemented **Option 2** in my `FlutterModule` project because it was easier to include pods from Flutter plugins. I had to create a custom build command based on the `BuildIOSFrameworkCommand` class to prevent the `_producePlugins` function from running. As a result, only `App.xcframework` was generated. In order to include Flutter plugins, their dependencies and `FlutterPluginRegistrant`, I combined this solution with [Option A](https://docs.flutter.dev/development/add-to-app/ios/project-setup#option-a---embed-with-cocoapods-and-the-flutter-sdk) from the add-to-app doc. Basically, for each `flutter_plugin` included a podspec from `.ios/.symlinks/plugins/<flutter_plugin>/ios/<flutter_plugin>.podspec` and for the `FlutterPluginRegistrant` a podspec from `.ios/Flutter/FlutterPluginRegistrant/FlutterPluginRegistrant.podspec`.  

---
So the main goal for me personally is to remove the custom build command (also `BuildIOSFrameworkCommand` class has a lot of private methods that I had to copy-paste to the subclass) and use a solution provided by Flutter. And I think it can be a good improvement in general. It also could help writing iOS version of https://github.com/flutter/website/issues/7775 because it currently seems impossible to easily share the same transitive platform dependency between Flutter module and iOS app

Please let me know if I misunderstood something and my use case can be achieved in some existing way

### PR DESCRIPTION
**Pull request by p-mazhnik**
A lot of details are written in the feature request: https://github.com/flutter/flutter/issues/114692.

tl;dr: Options B & C from the add2app iOS guide have a limitation (build error) in case the Flutter plugin and native iOS app have a shared dependency. We can use a workaround to avoid the issue, but in this case we don't need to build frameworks for plugins.

Closes https://github.com/flutter/flutter/issues/114692
Part of https://github.com/flutter/flutter/issues/130220

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by stuartmorgan** [comment]
Am I understanding correctly that the only benefit here is reducing build times?

If so, in isolation it's not clear to me that this is worth the complexity of adding and supporting. The flag seems rather difficult to use correctly, and it seems like the benefit of having it is minimal, so it seems like adding it as a supported output mode is something of a foot-gun.

**Comment by hyiso** [comment]
We also need this.

Currently we have one Flutter module  used in two iOS App, it has a number of plugins which transitively depend on many of native iOS Pods. 

We build this module twice with some difference in `--dart-define` passed to `flutter build ios-framework`.

During each build process, the `Building plugins` takes about 20 minutes because there are so many Pods being compiled, so the whole build process (running `flutter build ios-framework` twice) takes about 1 hour every time.

But we only need a new App.xcframework at the second build, so a flag like `--no-plugins` will decrease the build time a lot.

**Comment by hyiso** [comment]
@stuartmorgan Please reconsider this, reducing build time is something important when a project going large.

IMO, it is sad to ignore a helpful feature only for avoiding people use it wrong

**Comment by zhyky** [comment]
Similar problem, we are also facing the build time cost problem.

**Comment by christopherfujino** [comment]
Per discussion on discord, this is on @stuartmorgan TODO list.

**Comment by arthurmonteiroalvesmelo** [comment]
We need this too.

**Comment by eliasyishak** [comment]
@stuartmorgan friendly ping

**Comment by christopherfujino** [comment]
@stuartmorgan friendly ping

File: packages/flutter_tools/lib/src/commands/build_ios_framework.dart [hunk 1]
```diff
@@ -68,6 +68,11 @@ abstract class BuildFrameworkCommand extends BuildSubCommand {
       ..addFlag('cocoapods',
         help: 'Produce a Flutter.podspec instead of an engine Flutter.xcframework (recommended if host app uses CocoaPods).',
       )
+      ..addFlag('plugins',
+        defaultsTo: true,
+        help: 'Whether to produce frameworks for the plugins. '
+              'Consider using "--no-plugins" if plugin frameworks are not required.',
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_ios_framework.dart:74]: This second sentence doesn't seem to be adding anything; it's just saying to consider using the flag if you want the behavior of the flag. (It also sounds like people should pass it if they have no plugins, which isn't necessary or useful.)

I think this should instead discuss how this would need to be used. E.g., "This is intended for cases where plugins are already being built separately."

File: packages/flutter_tools/lib/src/commands/build_ios_framework.dart [hunk 2]
```diff
@@ -68,6 +68,11 @@ abstract class BuildFrameworkCommand extends BuildSubCommand {
...
+      )
       ..addFlag('static',
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_ios_framework.dart:76]: Passing both `--static` and `--no-plugins` should be an error, since that would be a mistake, and it's not clear what someone would expect it to do.

File: packages/flutter_tools/lib/src/commands/build_ios_framework.dart [hunk 3]
```diff
@@ -286,7 +291,7 @@ class BuildIOSFrameworkCommand extends BuildFrameworkCommand {
 
     globals.printStatus('Frameworks written to ${outputDirectory.path}.');
 
-    if (!project.isModule && hasPlugins(project)) {
+    if (!project.isModule && boolArg('plugins') && hasPlugins(project)) {
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_ios_framework.dart:294]: Why is this being gated on the `plugins` flag? This step should be basically free (so wouldn't impact build times, which as I understand it is the driver for this PR), and it's not at all clear to me that it wouldn't be needed in the use cases described here. Isn't the intended outcome that the app still has and uses plugins, they are just built separately? Building all the plugins via some other mechanism doesn't remove the need for a registrant.

**Review by stuartmorgan-g — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/lib/src/commands/build_ios_framework.dart (hunk 2 above)
Comment (p-mazhnik) [reply packages/flutter_tools/lib/src/commands/build_ios_framework.dart:76]: Added this to validations

**Review by stuartmorgan-g — APPROVED** [review]
LGTM. To @jmagman for a second review.

File: dev/devicelab/bin/tasks/build_ios_framework_module_test.dart [hunk 4]
```diff
@@ -816,6 +822,86 @@ Future<void> _testBuildMacOSFramework(Directory projectDir) async {
     outputPath,
     'GeneratedPluginRegistrant.swift',
   ));
+
+  section('Build frameworks without plugins');
+  await _testBuildFrameworksWithoutPlugins(projectDir, platform: 'macos');
+
+  section('check --static cannot be used with the --no-plugins flag');
+  await _testStaticAndNoPlugins(projectDir);
```
[the remark this answers is not in the document]
Comment (p-mazhnik) [reply dev/devicelab/bin/tasks/build_ios_framework_module_test.dart]: Good catch! `_testStaticAndNoPlugins` function run `build ios-framework` internally, that is why tests pass. `--static` is not supported for macos.

[the remark this answers is not in the document]
Comment (p-mazhnik) [reply dev/devicelab/bin/tasks/build_ios_framework_module_test.dart]: Fixed

**Review by jmagman — APPROVED** [review]
LGTM
